rholang/doc/rholangtut-0.2.md

```diff
@@ -0,0 +1,377 @@
+     7                 forwardFlag!(status) |
+     8                 status match { true => target!(msg) }
+     9             }
+    10         } |
+    11         for (_ <- kill; _ <- forwardFlag) {
+    12             forwardFlag!(false)
+    13         }
+    14     }
+    15 }
+
+2) We create a port to listen for method calls and a channel `forwardFlag` to store whether to forward messages.
+
+3) We return the channel on which clients send requests and the channel on which to send the kill signal.
+
+4) We set the initial state of `forwardFlag` to true.
+
+5-10) We read in an arbitrary message, get and replace the value of the flag.  If the flag is true, we forward the message to `target`.
+
+11-13) If a message is ever sent on the `kill` channel, we set `forwardFlag` to false.  The forwarder process then stops forwarding messages.
+
+### Composition
+
+By combining an attenuating forwarder with a revokable forwarder, we get both features:
+
+    new ret in {
+        MakeGetForwarder(target, ret) |
+        for (@pair <- ret) {
+            pair match [getOnly, kill] => {
+                MakeRevokableForwarder!(getOnly, *ret) |
+                for (revokableGetOnly <- ret) {
+                    // give away revokableGetOnly instead of target
+                    // hang onto kill for later revocation
+                }
+            }
+        }
+    }
+
+### Logging forwarder
+
+A logging forwarder can record all messages sent on a channel by echoing them to a second channel.
+
+    contract MakeLoggingForwarder(target, logger, ret) = {
+        new port in {
+            ret!(*port) |
+            contract port(msg) = {
+                target!(msg) |
+                logger!(msg)
+            }
+        }
+    }
+
+### Accountability
+
+Suppose Alice has a channel and would like to log Bob's access to it.  Bob would like to delegate the use of that channel to Carol and log her access.  Each party is free to construct their own logging forwarder around the channel they have received.  Alice will hold Bob responsible for whatever Carol does.
+
+### Sealing and unsealing
+
+    contract MakeSealerUnsealer(ret) =  {
+        new sealer, unsealer, ccRet in {
+            ret!(*sealer, *unsealer) |
+            MakeCoatCheck!(*ccRet) |
+            for (cc <- ccRet) {
+                contract sealer(@value, ret) = {
+                    cc!("new", value, *ret)
+                } |
+                contract unsealer(@ticket, ret) = {
+                    cc!("get", ticket, *ret)
+                }
+            }
+        }
+    }
+
+
+A sealer/unsealer pair gives the same functionality as public keys, but without cryptography. It's merely an attenuation of the coat check described above.  This design pattern can be used to sign something on a user's behalf.  In the Rholang blockchain tutorial, we'll see that it even works on the blockchain because there are no secrets to store, only unforgeable names to be kept inaccessible.
```

Comment:
functionality as public/private keys ... see that a sealer/unsealer pair even works as a signing/verification pair of keys on the blockchain because ...